### ISSUE TITLE
`one or more` rules consume trailing whitespace when exactly matched once

### ISSUE THREAD

**Issue by ejoebstl** [pest-parser/pest#396]
Given the following grammar: 

```
a = { "a" ~ b+ | a }
b = { "b" }
root = { SOI ~ a* ~ EOI }
```

Then, parsing the following text: 
```
a b

```
Will associate the trailing newline after `b` with the rule for `a`, instead of the the rule for `root`.

That seems to work properly with `zero-or-more` expressions, where the whitespace will be associated with the `root` rule. 

Also, it works properly for more than one `b`. The following example:
```
a b b

```
Puts the whitespace correctly inside root.

**Comment by ejoebstl** [comment]
The issue might be related to the following line, which inserts a concatenation between the first and the other repetitions. 

https://github.com/pest-parser/pest/blob/a0264c0633226708a086d298b50e59f9cbf8efef/meta/src/optimizer/unroller.rs#L18

**Comment by ejoebstl** [comment]
After looking through the code, I'd say this is because `Expr::RepOnce` is translated to a sequence. 

E.g. for pest, `b+` is equal to `b ~ b*`.

**Comment by ejoebstl** [comment]
Proposed fix in #397.

**Comment by dragostis** [comment]
Ah, finally. I knew this day was going to come. Unfortunately, this implementation bug is probably something that people make use of in their logic. Fixing this might actually break them, but I'm happy to give it a try regardless maybe there's no actual dependee that uses it.

I have a fix for 3.0, but I haven't had that much time to work on it.

**Comment by keks** [comment]
This just bit us. Are there still plans to resolve this? Is there evidence that people rely on this bug?

And lastly, is there a good workaround? The obvious idea would be to strip the white space off of everything pest returns, but I suspect there is a more elegant way, since otherwise there probably would be more people commenting here.

In our case we are having trouble to parse data in the format "x <- foo" (see [this page]).

Also, otherwise we are having a great time using pest, thank you!

cc @siccegge @chrisbrzuska

[this page]: https://codimd.faui2k.cloud/s/kjeothgO_#
